DM-40053: Add Datastore records to DatasetRef and use them in butler.get

andy-slac opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

andy-slac committed:
Add datastore records to DatasetRef

andy-slac committed:
Reimplement Butler.get with datastore records from registry.

There are few changes in Registry and Datastore classes that reduce use of
the bridge and opaque tables by datastores:
- New method `Datastore.get_opaque_table_definitions` returns opaque table
  definitoons which are used by Registry method `make_datastore_tables`
  to create those tables.
- `Registry.findDataset` has an option to retrieve datastore records
  and attach them to returned ref.
- `Datastore.get` can use the datastore records from the ref instead of
  using opaque tables.

Note that `Datastore.get` can still work with refs that don't have datastore
records, this is mostly to keep the unit tests running, a future commit will
remove that option and update all unit tests.

andy-slac committed:
Re-implement `Butler.put` with datastore records saved by Registry.

Adds `Datastore.put_new`, we want to keep old `Datastore.put` to keep
unit tests running (no unit tests for `put_new()` yet). `put_new()`
returns DatasetRef with datastore records, and new Registry method
is called by Butler to store those records in database. Better
approach would be for Registry to store datastore records in
`insertDatasets`/`_importDatasets` but that needs a bit of code
re-arrangement, will do that later.

andy-slac committed:
Remove dataset_id from `StoredDatastoreItemInfo` class.

This is to make it more similar to the `OpaqueTableValues` in DMTN-249
prototype. The `DatasetDatastoreRecords` type alias is similar to
`OpaqueTableBatch` from prototype. `StoredDatastoreItemInfo` is always
used with its corresponding `DatasetRef` or `DatasetId`, so keeping
dataset_id in each record is wasteful.

andy-slac committed:
Improve type checking fot generic datastore

andy-slac commented:
@timj, @TallJimbo, here is the current state of my branch, if you want to check it. Not so much code changes, but there were a lot of experiments along the way. Any comments are welcome!

timj commented:
I finally had a chance to look at this and given,

* We want datastores to have the option of not having to manage records at all.
* Each `DatasetRef` should carry its own datastore records.

then this looks like a reasonable approach. We do have to decide whether it's only ephemeral datastores that don't have to return records. Does SasquatchDatastore have to return empty records or only do that if `get()` is implemented?

andy-slac commented:
> Does SasquatchDatastore have to return empty records or only do that if `get()` is implemented?

I'm not familiar with SasquatchDatastore, but I can imagine that if it does not need datastore records stored in registry then it can return `None` or empty list/map for records.

andy-slac committed:
Add do-nothing implementation of new methods to remote registry.

This is only to suppress errors in the tests.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -85,6 +94,19 @@ class Event:
+
+    __slots__ = {"table_spec", "record_class"}
+    table_spec: ddl.TableSpec
+    record_class: type[StoredDatastoreItemInfo]
```
This is really a minor side issue, but I'd prefer not to assume anywhere that opaque tables are just for datastores, and I'd be great to cut down on the number of conversions (SQLAlchemy row <-> `StoredDatastoreItemInfo` <-> pydantic/json mapping).

So, could we put this `OpaqueTableDefinition` struct in `registry.interfaces.opaque`?

And could we make `record_class` just "some pydantic model type"?  Or maybe even just use bare `dict` here?   Dynamic class types aren't getting us any static-analysis advantages, and runtime checking is probably a waste of cycles (except at the JSON boundary).

andy-slac replied:
I think it should be OK to move `OpaqueTableDefinition` to `registry.interfaces`. I was worried that this may cause `core` to be (logically) dependent on `registry`, but we can do it as mypy-only dependency, as we do for a couple of other classes already. We do need then to replace `StoredDatastoreItemInfo` with registry-side type (or possibly use `dict`), I need to look at the code and figure out a reasonable option.

timj replied:
I don't like the idea of completely abandoning typing and using dict for arbitrary datastore records. The `StoredDatastoreItemInfo` dataclass definition is seemingly very close to defining the database schema directly through introspection.

andy-slac replied:
I would like to have typed datastore records, but it makes life a bit harder for registry. But it's probably unavoidable.

TallJimbo commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -413,11 +423,27 @@ def _get_stored_records_associated_with_refs(
+                        records_by_ref[ref.id].append(ref_record)
+
+        # If there were any refs without datastore records, check opaque table.
+        records = self._table.fetch(dataset_id=[ref.id for ref in refs_with_no_records])
```
This possibility is something I'd imagined disappearing from `Datastore` on DMTN-249: I was hoping we could make sure all possibly-relevant records were passed to a datastore whenever we'd get to code like this, such that if any were missing it'd represent repo corruption or logic bugs and hence the datastore should raise.  And that'd mean it wouldn't have a `self._table` at all.

Do you see this as transitional code, or something that'd stay for the long haul?

andy-slac replied:
This is indeed transitional, the final goal is to remove `self._table` completely. I just keep it because not all functionality is replaced yet and there are many unit tests to be updated.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -451,6 +447,53 @@ def put(self, inMemoryDataset: Any, ref: DatasetRef) -> None:
+            warnings.warn(f"Put of {ref} only succeeded in ephemeral databases", stacklevel=2)
+
+        if self._transaction is not None:
+            self._transaction.registerUndo("put", self.remove, ref)
```
This is a side issue, but `DatastoreTransaction` is definitely something I'd hoped we could remove after the DMTN-249 transition is complete and we require file artifacts to be successfully written or transferred _before_ we insert the corresponding rows into registry.

I'm hoping this is still here just because this PR doesn't try to get us all the way there?

andy-slac replied:
Yes, this PR is mostly to get a sense that new `DatasetRef` with all extra stuff in it looks reasonable, there are many things yet to update, including transaction handling by Datastore.

TallJimbo commented on python/lsst/daf/butler/registries/sql.py:
```diff
@@ -1341,6 +1350,48 @@ def queryDatasetAssociations(
                         timespan = None
                     yield DatasetAssociation(ref=ref, collection=collection_record.name, timespan=timespan)
 
+    def get_datastore_records(self, ref: DatasetRef) -> DatasetRef:
```
Eventually I think we'll really want this to be vectorized.

TallJimbo commented:
I think this is broadly consistent with what I think I'd have put in DMTN-249 if I hadn't been overzealous about optimizing the back-and-forth in `put` in client-server, but I need to go read that again (and maybe attempt to fix it) to be sure, at least if my question about datastores fetching missing records is resolved in favor of that being transitional.

andy-slac committed:
Rename OpaqueTableDefinition to DatastoreOpaqueTable

Old name was too generic, in reality this is very datastore-specific.
I do not think we need generic structure of the same kind, at least not
right now.

andy-slac committed:
Make datastore records private attribute of DatasetRef.

Jim suggested that we use new `OpaqueRecordSet` class to keep datastore
records, but after discussion we figured it would requre lots of additional
structure to support it. For now there is no clear need for that class and
we can continue using `StoredDatastoreItemInfo` in DatasetRefs, but want to
make records private to give us freedom to change it later.

andy-slac committed:
Make FileDatastore to ignore datastore records in some cases.

Datastore records in DatasetRef can become invalidated in some cases,
this is mostly true for our unit tests that do crazy things to test
consistency. This patch adds a flag parameter to some methods that
tells them to ignore ref-supplied datastore records.

andy-slac committed:
Revert implementation of `Butler.put` that used `Datastore.put_new`.

After trying to re-implement datastore unit tests I realized that the
`Datastore.put_new` and `Registry.store_datastore_records` combination
(used to implement `Butler.put`) is not very compatible with transaction
rollback system that we now have in place. So I decided to keep
`Butler.put` unchanged for now until we do something more drastic with
transactions. I keep `Datastore.put_new` and `Registry.store_datastore_records`
implementations, they are not used but may be useful in the future.

TallJimbo commented on python/lsst/daf/butler/datastore/_datastore.py:
```diff
@@ -1227,6 +1263,19 @@ def set_retrieve_dataset_type_method(self, method: Callable[[str], DatasetType |
         """
         pass
 
+    @abstractmethod
```
Should make this a `property` or add `get_` to the name.

andy-slac replied:
I'll add `get_`, I'm not sure if property is the correct thing for this.

TallJimbo commented on python/lsst/daf/butler/datastore/_datastore.py:
```diff
@@ -537,6 +553,26 @@ def put(self, inMemoryDataset: Any, datasetRef: DatasetRef) -> None:
         raise NotImplementedError("Must be implemented by subclass")
 
+    @abstractmethod
+    def put_new(self, inMemoryDataset: Any, datasetRef: DatasetRef) -> Mapping[str, DatasetRef]:
```
```suggestion
    def put_new(self, in_memory_dataset: Any, dataset_ref: DatasetRef) -> Mapping[str, DatasetRef]:
```

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1004,6 +1008,7 @@ def _findDatasetRef(
         collections: Any = None,
         predict: bool = False,
         run: str | None = None,
+        datastore_records: bool = False,
```
Needs a doc entry.

TallJimbo commented on python/lsst/daf/butler/_dataset_ref.py:
```diff
@@ -176,6 +182,10 @@ def makeDatasetId(
 
+# Serialized representation of StoredDatastoreItemInfo collection, first item
+# is the record class name.
+_DatastoreRecords: TypeAlias = tuple[str, list[Mapping[str, Any]]]
```
If this is just for serialization we may want to put that in the name.

What breaks if we don't serialize these records at all?  I am a bit worried that if we start including them in `SerializedDatasetRef` by default we might blow up the size of `QuantumGraph` without solving any problems (yet).

andy-slac replied:
You are right, we don't have an immediate use for this, so it may be better to drop it completely. I guess DM-41115 will take care of datastore record serialization for future QuantumGraph?

TallJimbo replied:
> I guess DM-41115 will take care of datastore record serialization for future QuantumGraph?

Yes, that's how I'm hoping things will go.

andy-slac replied:
Oops, I forgot to undo this part, will make another ticket.

TallJimbo commented on python/lsst/daf/butler/datastore/generic_base.py:
```diff
@@ -32,136 +32,35 @@
     def _post_process_get(
         self,
-        inMemoryDataset: Any,
+        inMemoryDataset: object,
```
Any particular reason for `object` vs. `Any` here?

andy-slac replied:
This is mostly personal preference, I think mypy can catch more potential issues with `object` as `Any` disables all checks. Some docstrings already specified `object` for parameters, this is just restoring documented behavior 🙂

andy-slac replied:
I can revert to `Any` if consistency is important.

TallJimbo replied:
`object` is fine.  I was mostly curious for my own education - I knew mypy was a bit more permissive with `Any` but I didn't have any concrete examples of how it was different in my mind.

andy-slac replied:
It's probably mostly about accidental corner cases, but here is a trivial example to show differences:
```
from typing import Any

class Butler():
    def get_obj(self) -> object:
        return 42

    def get_any(self) -> Any:
        return 42

b = Butler()

x = [42]
x = b.get_obj()  # will be mypy error
x = b.get_any()  # mypy is OK
l = len(x)  # mypy is still OK, as x is Any here

y = b.get_obj()
y.something()  # mypy error

z = b.get_any()
z.something()  # mypy is OK
```

TallJimbo replied:
Ah, interesting; `object` is _much_ less permissive than I thought - it seems to be an object you can't do anything with (except basic stuff like equality tests), while `Any` permits anything (including implicit downcasts).

TallJimbo commented on python/lsst/daf/butler/datastore/record_data.py:
```diff
@@ -63,15 +61,17 @@ class SerializedDatastoreRecordData(_BaseModelCompat):
+    records: Mapping[str, Mapping[str, Mapping[str, list[_Record]]]]
+    """List of records indexed by record class name, dataset ID (encoded as
+    str, because JSON), and opaque table name.
+    """
```
It's too bad we can't use opaque table name to get the record class name instead of storing them both, but I see this is preexisting (and just used in `Quantum` and `QuantumBackedButler`, right?) so maybe this is not the time to try to fix that.

andy-slac replied:
The mapping from opaque table name to record class would have to exist somewhere when de-serialization happens. I'm not sure what would be the right place for it.

TallJimbo replied:
I'm fine with leaving this.  DM-41115 may also change this completely anyway.

TallJimbo commented on python/lsst/daf/butler/datastore/stored_file_info.py:
```diff
@@ -87,13 +89,14 @@ def from_record(cls: type[StoredDatastoreItemInfo], record: Mapping[str, Any]) -
-    def to_record(self) -> dict[str, Any]:
-        """Convert record contents to a dictionary."""
-        raise NotImplementedError()
+    def to_record(self, **kwargs: Any) -> dict[str, Any]:
```
This may not be time to change it, but I'm not a fan of the name of this; we call these things "records" before we convert them to `dict`, so there's nothing here that implies conversion to `dict`.

andy-slac replied:
Yes, the naming is confusing, we call instances infos or records in different places. I can rename this method (and `from_record`, `to_records`, `from_records`) to `as_dict`, etc. @timj, do you have preference for keeping this naming or some other names?

TallJimbo replied:
Renaming `StoredDatastoreItemInfo` -> `DatastoreRecord`  and `StoredFileInfo` -> `FileDatastoreRecord` would be ideal but maybe not a good idea (especially on this ticket).

I do think `to_dict` is better than `to_records` (same with `from_`), but I think we may well want these to be native pydantic models, which would mean we'd just call `model_dump` instead of having another method.

andy-slac replied:
OK, I'll leave these as they are for now, and wait for big refactoring.

TallJimbo commented on python/lsst/daf/butler/datastore/stored_file_info.py:
```diff
@@ -102,12 +105,56 @@ def update(self, **kwargs: Any) -> StoredDatastoreItemInfo:
     """Datastore-private metadata associated with a Datastore file."""
 
-    __slots__ = {"formatter", "path", "storageClass", "component", "checksum", "file_size", "dataset_id"}
+    __slots__ = {"formatter", "path", "storageClass", "component", "checksum", "file_size"}
```
We can use `dataclass(..., slots=True)` now to avoid spelling these out.

TallJimbo commented on python/lsst/daf/butler/datastore/stored_file_info.py:
```diff
@@ -239,14 +280,15 @@ def from_record(cls: type[StoredFileInfo], record: Mapping[str, Any]) -> StoredF
         component = record["component"] if (record["component"] and record["component"] != NULLSTR) else None
 
+        # UUID may be converted to string, e.g. in round-trip through JSON,
+        # convert it back to UUID.
```
This comment doesn't seem to be describing any code around it.

andy-slac replied:
Leftover for previous commits, I dropped dataset_id from the class, will remove this comment.

TallJimbo commented on python/lsst/daf/butler/datastores/inMemoryDatastore.py:
```diff
@@ -62,7 +60,7 @@ class StoredMemoryItemInfo(StoredDatastoreItemInfo):
     """
 
-    __slots__ = {"timestamp", "storageClass", "parentID", "dataset_id"}
+    __slots__ = {"timestamp", "storageClass", "parentID"}
```
Another opportunity for passing `slots=True` to the `dataclass` decorator.

TallJimbo commented on python/lsst/daf/butler/registries/sql.py:
```diff
@@ -219,6 +221,10 @@ def __init__(self, database: Database, defaults: RegistryDefaults, managers: Reg
 
+        # TODO: This is currently initialized by `make_datastore_tables`,
+        # eventually we'll need to do it during construction.
+        self._datastore_record_classes: Mapping[str, type[StoredDatastoreItemInfo]] = {}
```
Are keys here datastore names or opaque table names?

I also wonder if it'd be better to just have a mapping of `DatastoreOpaqueTable`.

andy-slac replied:
Keys are opaque table names, I'll update comment. Registry itself does not need `TableSpec` part of `DatastoreOpaqueTable`, depending on how we initialize this structure in the future it may be better not to require `TableSpec` at that point.

TallJimbo commented on python/lsst/daf/butler/registry/_butler_registry.py:
```diff
@@ -216,3 +219,53 @@ def getDatastoreBridgeManager(self) -> DatastoreRegistryBridgeManager:
+        Notes
+        -----
+        If this method is called with the dataset ref that is not known to the
+        registry then the reference with an empty set of records is returned.
```
I'm a bit surprised by this behavior; generally in methods that add records to refs, I think we want:

- `DatasetRef._datastore_records = None`: this ref might have records, but we haven't tried to "expand" the ref to get them yet;
- `DatasetRef._datastore_Records = {}`: this ref is known to the registry but is not in any datastore and hence has no datastore records;
- (raise an exception): this ref is not known to the registry at all.

But if it's expensive or otherwise inconvenient to maintain that invariant maybe it's not worth it.

andy-slac replied:
The complication here is InMemoryDatastore which does not have any records. I want to avoid raising an exception in that case because Butler would need then to figure out what to do with it. This is not ideal of course as we now depend on Butler+Datastore doing correct thing when the records are empty. The alternative is to do an additional check for dataset_id in that case, which is one more query on a large table. I think I want to see additional use cases for this method to decide if this needs change.

TallJimbo replied:
If the answer is "because InMemoryDatastore" then let's just revisit it when that's dead and gone.

TallJimbo reviewed (approved):
I've left a lot of little comments and questions but I have no big-picture concerns; it's all a step in the right direction.

FWIW, your `put_new` does not match the one I have been prototyping with transaction support over on DMTN-249 anymore, and I think it'll probably end up changing again when we do transactions and/or client-server `put`, but my prototyping doesn't include `ChainedDatastore` so I'm not confident it will work as-is either.

andy-slac committed:
Apply review suggestions

Co-authored-by: Jim Bosch <jbosch@astro.princeton.edu>

andy-slac committed:
Use slots flag with datastore records dataclasses
